### PR TITLE
Trade money

### PR DESCRIPTION
**Pull request by cpranav93**
Adds the function for trading money between agents. Two methods are implemented:
- weighted_transfer - for sharing of wealth to all neighbors based on a edge weight function
- singular_transfer - for randomly choosing one neighbour by subsampling the graph and transfering wealth

### CODE REVIEW

**Comment by cpranav93** [comment]
There are multiple files in this PR simply because cookie cutter and dgl_framework are yet to merged. It should simplify after the merge.

**Comment by vmgaribay** [comment]
Looks good, thanks for working on both options. Currently there is no threshold in place for a trade to occur, right? I just wanted to make sure I understood the code correctly.

**Comment by cpranav93** [comment]
Yes. Currently when a single neighbor is chosen, they always get some wealth (lambda * k). Does this need to be randomized? On whether or not they get any money?

File: dgl_ptm/dgl_ptm/trade_money.py [hunk 1]
```diff
@@ -1,7 +1,75 @@
-#!/usr/bin/env python
-# coding: utf-8
+import os
+import torch 
+import dgl 
+import dgl.function as fn
 
-# trade_money - Trades money between the different connected agents
+os.environ["DGLBACKEND"] = "pytorch"
```
Comment (meiertgrootes) [inline dgl_ptm/dgl_ptm/trade_money.py:6]: I understand that we need it here if we want to simply call this stand-alone for the time being, but this is sth. that would be set in the main function as it sets global environment variables at the OS level, if we leave it in here we could get conflicts somewhere down the line

Comment (cpranav93) [reply dgl_ptm/dgl_ptm/trade_money.py]: Yeah this makes sense. I have removed the line now. We can add it in main based on the user chosen backend.

File: dgl_ptm/dgl_ptm/trade_money.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+                         Method needs to be either 'weighted_transfer' or 'singular_transfer'")
+    
+def _weighted_transfer(agent_graph):
+    """
+        Weighted transfer of wealth from each agent (node) to every connected 
+        neighbour agent based on pre-defined edge weights.
+    """
+
+    # Sum all incoming weights
+    agent_graph.ndata['tot_wgt'] = torch.zeros(agent_graph.num_nodes(),1)
```
Comment (vanlankveldthijs) [inline dgl_ptm/dgl_ptm/trade_money.py:47]: I prefer more verbose variable names; I don't really see the point for shortening things like "total" to "tot" but this is mostly harmless as far as I'm concerned.

It becomes a problem for terms like "per" (per what?), "disp" (disparate? display? distribution P value?) or "trfr".
We don't gain anything from writing shorter lines of code if this decreases human readability.

Things like "wgt" and "msg" that are reused almost everywhere could be exampt from this verbosity.

File: dgl_ptm/dgl_ptm/trade_money.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    agent_graph.update_all(fn.u_add_e('tot_wgt','w','tot_wgt'), fn.sum('tot_wgt', 'tot_wgt'))
```
Comment (vanlankveldthijs) [inline dgl_ptm/dgl_ptm/trade_money.py:48]: Didn't we discuss adding the "_msg" postfix to the center two variables to make explicit that they are different from the outer two?

That is: "agent_graph.update_all(fn.u_add_e('tot_wgt','w','tot_wgt_msg'), fn.sum('tot_wgt_msg', 'tot_wgt'))"

The same holds for the other instances of message passing.

File: dgl_ptm/dgl_ptm/trade_money.py (hunk 2 above)
Comment (cpranav93) [reply dgl_ptm/dgl_ptm/trade_money.py:47]: @vanlankveldthijs, the intention was to rename the variables as per your naming convention but at that time, I could not find your notebook online. Is it pushed now? Could you maybe direct me to it? 

I currently have this in the file:
TODO: Rename variables as per Thijs' updates on notebook

to ensure we remember to rename it verbosely. :D

Comment (vanlankveldthijs) [reply dgl_ptm/dgl_ptm/trade_money.py:47]: I never got round to finishing that notebook at the end of the sprint and now I cannot find it. Oops (:-/).

All I remember is:
* basically spelling out the whole words for most shortened terms (e.g. "tot_wgt" -> "total_weight"),
* adding the _msg prefix to messages (e.g. "g.update_all(fn.u_add_e('tot_wgt','w','tot_wgt'), fn.sum('tot_wgt', 'tot_wgt'))" -> g.update_all(fn.u_add_e('tot_wgt','w','tot_wgt_msg'), fn.sum('tot_wgt_msg', 'tot_wgt')))
* In certain cases adding a bit more to variables of complex concepts (e.g. "weight_ratio_at_source" for the variable on an edge that specifies how big the ratio of the edge's weight is relative to all the edge weights with the same source)

**Comment by meiertgrootes** [comment]
merging as agreed. Pending a cleaning of naming in full code base

**Pull request merged**